### DISCUSSION TITLE
Idea: watch mode for build

### DISCUSSION THREAD

**Discussion by Satont** [withastro/roadmap#273] [Proposal]
Use something like `astro build --watch`, as it done in vite `vite build --watch`. 

So, files emitted on every change. 

Might be helpful for node ssr.

**Comment by smnbbrv** [comment]
Upvoting as currently one needs to make the following to make it work:

package.json

```
{
  "name": "@example/basics",
  "type": "module",
  "version": "0.0.1",
  "private": true,
  "scripts": {
    "start": "run-p start:src start:srv",
    "start:src": "nodemon --config nodemon.src.json --exec 'astro build'",
    "start:srv": "sleep 2 && nodemon --config nodemon.srv.json srv/main.ts",
    "build": "astro build",
    "preview": "astro preview",
    "astro": "astro"
  },
  "dependencies": {
    "@astrojs/node": "^2.0.0",
    "astro": "^1.5.0",
    "express": "^4.18.2"
  },
  "devDependencies": {
    "@types/express": "^4.17.14",
    "nodemon": "^2.0.20",
    "ts-node": "^10.9.1"
  }
}
```

nodemon.src.json

```
{
    "verbose": false,
    "watch": [
        "src/**/*"
    ],
    "ext": "astro,ts"
}
```

nodemon.srv.json

```
{
    "verbose": false,
    "watch": [
        "dist/server/**/*.mjs",
        "srv/**/*.ts"
    ],
    "ext": "ts,json,mjs",
    "execMap": {
        "ts": "ts-node-esm"
    }
}
```

That is really, well, not that clean setup, it works, but I assume it would get way slower in the project with a decent amount of components

**Comment by coryvirok** [comment]
Looks like this idea is in the same direction as https://github.com/withastro/rfcs/discussions/340 and https://github.com/withastro/rfcs/discussions/344. 

For this issue, (adding a watch mode for build) I'm assuming you want this because you need the production build in order to run your server. Presumably because you are running an Express app with Astro building in middleware mode. If that's the case, I think a better solution than adding a watch flag to build would be to **expose the Astro dev server as a middleware** that you could add to your Express app. 

I'm pretty new to Astro but this seems like a growing need since `astro dev` is useless if you deploy Astro as a middleware to hook into a larger Express app.

**Reply by awwong1** [reply]
When using the [@astrojs/cloudflare](https://docs.astro.build/en/guides/integrations-guide/cloudflare/) adapter, proxying the dev server is inadequate because the `.astro` files are not running in wrangler and are [missing the Cloudflare workers context](https://discord.com/channels/830184174198718474/1002802338538463313/1011642685406326814).

i.e. `npx wrangler pages dev --proxy 3000 -- npx astro dev` is currently not sufficient to do local preview of the web app.

We could however have a build watch, then rely on `npx wrangler pages dev ./dist --live-reload` to handle changes that occur within `./dist`.

**Reply by alexanderniebuhr** [reply]
[@astrojs/cloudflare](https://docs.astro.build/en/guides/integrations-guide/cloudflare/) is my main use-case as-well. What is the go-to workaround for wrangler?

**Comment by simontong** [comment]
Has there been any movement on this?

**Reply by alexanderniebuhr** [reply]
Yes there is, if you would like to help, please see: 
https://github.com/withastro/roadmap/discussions/590#discussioncomment-6924426

**Comment by alexanderniebuhr** [comment]
We've added support for `local` bindings in `astro dev`. Would appreciate testing & feedback!

https://docs.astro.build/en/guides/integrations-guide/cloudflare/#runtime

**Comment by Always-Self-Hosted** [comment]
For some extra context, I want to use astro as a static site generator, and i use bun for my tooling and to serve the static files. So ideally what i want is for this to work:

```
"scripts": {
    "dev": "bun run build && bun --watch dist/*.html",
    "build": "bun x --bun astro build --watch",
  }
  ```
  
  Which will essentially rebuild my dist html files on file save, which then bun also restarts its process when the files have rebuilt.
